mm2src/coins/z_coin.rs

```diff
@@ -1,94 +1,564 @@
+    };
+
+    let (utxo_weak, z_fields_weak) = coin.into_weak_parts();
+
+    let zero_root = Some(H256Json::default());
+    while let Some(coin) = ZCoin::from_weak_parts(&utxo_weak, &z_fields_weak) {
+        coin.z_fields.sapling_state_synced.store(false, AtomicOrdering::Relaxed);
+        let current_block = match coin.rpc_client().get_block_count().compat().await {
+            Ok(b) => b,
+            Err(e) => {
+                log::error!("Error {} on getting block count", e);
+                Timer::sleep(10.).await;
+                continue;
+            },
+        };
+
+        let native_client = match coin.rpc_client() {
+            UtxoRpcClientEnum::Native(n) => n,
+            _ => unimplemented!("Implemented only for native client"),
+        };
+        while processed_height as u64 <= current_block {
+            let block = match native_client.get_block_by_height(processed_height as u64).await {
+                Ok(b) => b,
+                Err(e) => {
+                    log::error!("Error {} on getting block", e);
+                    Timer::sleep(1.).await;
+                    continue;
+                },
+            };
+            let current_sapling_root = current_tree.root();
+            let mut root_bytes = [0u8; 32];
+            current_sapling_root
+                .write(&mut root_bytes as &mut [u8])
+                .expect("Root len is 32 bytes");
+
+            let current_sapling_root = Some(H256::from(root_bytes).reversed().into());
+            if current_sapling_root != block.final_sapling_root && block.final_sapling_root != zero_root {
+                let prev_tree_state = current_tree.clone();
+                let mut cmus = Vec::new();
+                for hash in block.tx {
+                    let tx = native_client
+                        .get_transaction_bytes(hash)
+                        .compat()
+                        .await
+                        .expect("Panic here to avoid storing invalid tree state to the DB");
+                    let tx: UtxoTx = deserialize(tx.as_slice()).expect("Panic here to avoid invalid tree state");
+                    for output in tx.shielded_outputs {
+                        current_tree
+                            .append(Node::new(output.cmu.take()))
+                            .expect("Commitment tree not full");
+                        cmus.push(output.cmu);
+                    }
+                }
+
+                let state_to_insert = SaplingBlockState {
+                    height: processed_height + 1,
+                    prev_tree_state,
+                    cmus,
+                };
+                insert_block_state(&coin.sqlite_conn(), state_to_insert).expect("Insertion should not fail");
+            }
+            processed_height += 1;
+        }
+        coin.z_fields.sapling_state_synced.store(true, AtomicOrdering::Relaxed);
+        drop(coin);
+        Timer::sleep(10.).await;
```

Comment:
I think that I tested this and the transaction was successfully generated and broadcast even after a new block appeared. I will recheck this.

> the user has spent his outputs

Could you please provide more details on this? What situation do you assume?

Reply:
I meant a case like the following:
* the user withdraw his founds;
* the transaction is mined;
* the user tries to withdraw before `sapling_state_cache_loop` fetched the last block.
So the new transaction may include already spent inputs.

Reply:
> What situation do you assume?

It's not a case because I realized that `get_unspent_witness` waits for 10s until the block with an unspent output is fetched.